File: two-pointers/easy/valid-palindrome/sol.js
// https://leetcode.com/problems/valid-palindrome/

/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function (s) {
  let l = 0,
    r = s.length - 1;

  while (l < r) {
    while (l < r && !an(s[l])) {
      l++;
    }
    while (r > l && !an(s[r])) {
      r--;
    }
    if (s[l].toLowerCase() !== s[r].toLowerCase()) {
      return false;
    }
    l++;
    r--;
  }

  return true;

  function an(c) {
    return (
      ("a" <= c && c <= "z") || ("A" <= c && c <= "Z") || ("0" <= c && c <= "9")
    );
  }
};

File: two-pointers/easy/valid-palindrome/sol.cs
// https://leetcode.com/problems/valid-palindrome/

public class Solution
{
  public bool IsPalindrome(string s)
  {
    int l = 0, r = s.Length - 1;

    while (l < r)
    {
      while (l < r && !an(s[l]))
      {
        l++;
      }
      while (r > l && !an(s[r]))
      {
        r--;
      }
      if (Char.ToLower(s[l]) != Char.ToLower(s[r]))
      {
        return false;
      }
      l++; r--;
    }

    return true;

    static bool an(char c)
    {
      return (
          ('a' <= c && c <= 'z')
          || ('A' <= c && c <= 'Z')
          || ('0' <= c && c <= '9')
      );
    }
  }
}